src/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestService.cs

```diff
@@ -29,66 +31,90 @@ public TestService(string serviceName)
             // We cannot easily test these so disable the events
             this.CanHandleSessionChangeEvent = false;
             this.CanHandlePowerEvent = false;
-        }
 
-        public static string GetLogPath(string serviceName)
-        {
-            return typeof(TestService).Assembly.Location + "." + serviceName + ".log";
+            this.Server = new NamedPipeServerStream(serviceName);
+            this.Server.WaitForConnectionAsync();
         }
 
+        public NamedPipeServerStream Server { get; set; }
+
         protected override void OnContinue()
         {
-            WriteLog(nameof(OnContinue));
             base.OnContinue();
+            WriteStream(PipeMessageByteCode.Continue);
         }
 
         protected override void OnCustomCommand(int command)
         {
-            WriteLog(nameof(OnCustomCommand) + " command=" + command);
             base.OnCustomCommand(command);
+            WriteStream(PipeMessageByteCode.OnCustomCommand, command);
         }
 
         protected override void OnPause()
         {
-            WriteLog(nameof(OnPause));
             base.OnPause();
+            WriteStream(PipeMessageByteCode.Pause);
         }
 
         protected override void OnSessionChange(SessionChangeDescription changeDescription)
         {
-            WriteLog(nameof(OnSessionChange) + " change=" + changeDescription.ToString());
             base.OnSessionChange(changeDescription);
         }
 
         protected override bool OnPowerEvent(PowerBroadcastStatus powerStatus)
         {
-            WriteLog(nameof(OnPowerEvent) + " status=" + powerStatus.ToString());
             return base.OnPowerEvent(powerStatus);
         }
 
         protected override void OnShutdown()
         {
-            WriteLog(nameof(OnShutdown));
             base.OnShutdown();
         }
 
         protected override void OnStart(string[] args)
         {
-            File.Delete(GetLogPath(ServiceName));
-
-            WriteLog(nameof(OnStart) + " args=" + string.Join(",", args));
             base.OnStart(args);
+            if (args.Length == 4 && args[0] == "StartWithArguments")
+            {
+                Debug.Assert(args[1] == "a");
+                Debug.Assert(args[2] == "b");
+                Debug.Assert(args[3] == "c");
+                Server.WaitForConnectionAsync();
```

Comment:
That feels strange as it ends up forcing another connect, what value does that provide?

Reply:
on doing controller.stop() ..the server is getting disconnected..so in order for server to listen..we have to start the server again

Reply:
Yes, but here it is async anyway so there is no guarantee that it is connected by the time that you get to write to it. What is likely happening is that the one that you triggered at the constructor is already completed when you get here.

Reply:
You will still require the connect on the client side but this call here is not necessary.

Reply:
yeah you are write that it may not be connected when I am writing to the stream. We can change that to synchronous wait. The connection triggered by the constructor has been completed and even destroyed. eg Controller.stop() checks if the constructor connection is completed, if it is completed it proceeds and at the disconnects the connection.